### PR TITLE
hcm: bugfix - request hangs where 1st filter adds body, and 2nd filter waits for it

### PR DESCRIPTION
**Pull request by yuval-k**
Commit Message:
Fix a bug where a request without a body hangs where a filter adds a body, and a subsequent filter stops the iteration until decodeData.

Additional Description:
We found this when one of our customers tried to use our transformation filter to add a body to a header-only request, and then sent it to AWS lambda.

Risk Level: Small-medium
Testing: Added unit test that fails before the change, and passes now.
Docs Changes: N/A
Release Notes: N/A

Fixes https://github.com/solo-io/gloo/issues/3019

### CODE REVIEW

**Comment by yuval-k** [comment]
it looks like tsan failed because of a transient github error:
```
2020-05-18T22:07:50.2478846Z 2020/05/18 22:07:50 could not resolve the version 'latest' to an actual version number: could not get releases from github.com/bazelbuild/bazel: could not download list of Bazel releases from github.com/bazelbuild: unexpected status code while reading https://api.github.com/repos/bazelbuild/bazel/releases: 504
2020-05-18T22:07:50.2536470Z ci/do_ci.sh: line 32: pushd: bazel-testlogs: No such file or directory
```

**Comment by yuval-k** [comment]
coverage fails (`Code coverage 96.7 is lower than limit of 97.0`), but the weird part is that the coverage drop is in code that i didn't touch.
comparing the coverage in this commit and in the master commit that comes before it, i see the biggest drop in coverage in `source/server` package.
for reference:
mine `source/server` coverage report:
https://346874-65214191-gh.circle-artifacts.com/0/coverage/source/server/index.html
the commit before me coverage report:
https://346805-65214191-gh.circle-artifacts.com/0/coverage/source/server/index.html

specifically the coverage for `hot_restarting_base.cc` tanked. was this observed before?

**Comment by jmarantz** [comment]
RE covrerage: one thing I saw recently was a PR that was passing coverage. Then I *added* a test, and it failed coverage (!)

The reason I think was there was an operational failure in the script to collect some of the coverage data from the run. I didn't look at why. But I re-ran the coverage test and it passed.

@lizan can we make the coverage script surface such operational failures in a more obvious way?

/retest

File: test/integration/filters/wait_for_whole_request.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include "envoy/server/filter_config.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/extensions/filters/http/common/empty_http_filter_config.h"
+#include "test/integration/filters/common.h"
+
+namespace Envoy {
+
+class WaitForWholeRequestStreamFilter : public Http::PassThroughFilter {
```
Comment (mattklein123) [inline test/integration/filters/wait_for_whole_request.cc:14]: nit: I think this is WaitForWholeRequestAndResponseStreamFilter or similar? (Same for other names like this.) Can you also add a small comment up top for what this does like you did with the other one?

**Pull request merged**